docs/source/developer/debugging.rst

```diff
@@ -0,0 +1,130 @@
+defined in it.
+
+The CPython source distribution includes a suppressions file, in the file
+``Misc/valgrind-python.supp``. Using this file prevents a lot of spurious errors
+generated by Python's memory allocation implementation. Additionally, the Numba
+repository includes a suppressions file in ``contrib/valgrind-numba.supp``.
+
+To run the Python interpreter under Memcheck with both suppressions
+files, it is invoked with the following command::
+
+   valgrind --tool=memcheck \
+            --suppressions=${CPYTHON_SRC_DIR}/Misc/valgrind-python.supp \
+            --suppressions=${NUMBA_SRC_DIR}/contrib/valgrind-numba.supp \
+            python ${PYTHON_ARGS}
+
+where ``${CPYTHON_SRC_DIR}`` is set to the location of the CPython source
+distribution, ``${NUMBA_SRC_DIR}`` is the location of the Numba source dir, and
+``${PYTHON_ARGS}`` are the arguments to the Python interpreter.
+
+If there are errors, then messages describing them will be printed to standard
+error. An example of an error is::
+
+   ==77113==    at 0x24169A: PyLong_FromLong (longobject.c:251)
+   ==77113==    by 0x241881: striter_next (bytesobject.c:3084)
+   ==77113==    by 0x2D3C95: _PyEval_EvalFrameDefault (ceval.c:2809)
+   ==77113==    by 0x21B499: _PyEval_EvalCodeWithName (ceval.c:3930)
+   ==77113==    by 0x26B436: _PyFunction_FastCallKeywords (call.c:433)
+   ==77113==    by 0x2D3605: call_function (ceval.c:4616)
+   ==77113==    by 0x2D3605: _PyEval_EvalFrameDefault (ceval.c:3124)
+   ==77113==    by 0x21B977: _PyEval_EvalCodeWithName (ceval.c:3930)
+   ==77113==    by 0x21C2A4: _PyFunction_FastCallDict (call.c:376)
+   ==77113==    by 0x2D5129: do_call_core (ceval.c:4645)
+   ==77113==    by 0x2D5129: _PyEval_EvalFrameDefault (ceval.c:3191)
+   ==77113==    by 0x21B499: _PyEval_EvalCodeWithName (ceval.c:3930)
+   ==77113==    by 0x26B436: _PyFunction_FastCallKeywords (call.c:433)
+   ==77113==    by 0x2D46DA: call_function (ceval.c:4616)
+   ==77113==    by 0x2D46DA: _PyEval_EvalFrameDefault (ceval.c:3139)
+   ==77113==
+   ==77113== Use of uninitialised value of size 8
+
+The traceback provided only outlines the C call stack, which can make it
+difficult to determine what the Python interpreter was doing at the time of the
+error. One can learn more about the state of the stack by looking at the
+backtrace in GDB. Launch ``valgrind`` with an additional argument,
+``--vgdb-error=0`` and attach to the process using GDB as instructed by the
+output. Once an error is encountered, GDB will stop at the error and the stack
+can be inspected.
+
+GDB does provide support for backtracing through the Python stack, but this
+requires symbols which may not be easily available in your Python distribution.
+In this case, it is still possible to determine some information about what was
+happening in Python, but this depends on examining the backtrace closely. For
+example, in a backtrace corresponding to the above error, we see items in the
+backtrace such as:
```

Comment:
Perhaps ` we see items such as` as the `backtrace` is referenced previously?

Reply:
Yeah, that would be less redundant and also less redundant.

Reply:
Duplicate reference to backtrace removed.